### ISSUE TITLE
Proposal for Package Ranking

### ISSUE THREAD

**Issue by carlisia** [golang/gddo#320]
## Rank by Objective Quality Standards
A mechanism to transparently and objectively rank projects by quality level.

We're proposing to avoid the use of a user-based rating or feedback system. It would require the creation and maintenance of accounts and would be subject to abuse. Instead, we propose using a published set of project quality standards that are not subject to simple manipulation:

- Percent of code coverage achieved by unit tests
- Presence of package and type documentation for public types
- Number of forks and downloads on repository. 
- Measure of recent activity on the project (commits in the last X days)
- Number of includes in other public projects. 
- Number of pageviews for that project on GoDoc.org
- [other criteria to be established by the community]

This could use (via API or similar) the system implemented by [GoReportCard](http://goreportcard.com/report/go-sql-driver/mysql), for example:
![image](https://cloud.githubusercontent.com/assets/16508/10084126/a30635c6-62b6-11e5-9ac9-9033383ce600.png)

The results of the ranking score would be used as the primary sort mechanism when browsing packages by search or by category. The information would also be linked/displayed at the top of each documentation page on Godoc.org.

**References:**
Examples of quality assessement tools: https://medium.com/@jgautheron/quality-pipeline-for-go-projects-497e34d6567

## Include Ranking in Display
Once the ranking system is in place, include a summary of the score (and link to view details/explanation) on both the documentation pages and in a second column on search results.

---
### Contributors to this proposal

@carlisia
@gdey
@rafaeljusto

**Comment by garyburd** [comment]
What are your goals for search ranking and how do the listed metrics support these goals?

One goal might be to help developers find the best package in a some domain.  Ranking by number of imports supports this goal because each import is an indication that a developer found the package useful.  Ranking by recent activity does not necessarily support this goal. Activity on a high quality mature package can be low. Activity on a buggy package can be high.

**Comment by adg** [comment]
I'm in favor of adding more signal to godoc.org's ranking algorithm. As such, I'm generally in favor of this proposal.

Let me comment on a few of the proposed metrics, and how they might be computed:

> Percent of code coverage achieved by unit tests

This requires integration with a continuous integration system. As such, it's probably the hardest one to gather data for. Defer til last. (Also, this is easily gamed; just add a function 100k lines long and write one test to invoke it.)

> Presence of package and type documentation for public types

Sounds good, if we're not doing this already (can't recall). I implemented this in a [separate project](https://github.com/nf/godash) a long time ago.

> Number of forks and downloads on repository.
> Measure of recent activity on the project (commits in the last X days)

How do we measure repository downloads? GitHub stars seem like good signals. Forks are ambiguous (@garyburd raises some interesting questions).

> Number of includes in other public projects.

We already have this data. Seems like a no-brainer. 

> Number of pageviews for that project on GoDoc.org

We're not currently gathering this data, but it's probably worth doing.

--

When we decide to embark on implementing any of these specific metrics, please create a separate issue for that particular metric so that we can nail down the design before implementation.

**Comment by jbuberel** [comment]
@garyburd That's a good point about low activity on an established, high-quality project. Do you agree with @adg's suggestion that GitHub starts are a reasonable proxy for "this is a good project"?

One of the odd things about Github stars is that they're cumulative, with no decay function. You can star a project at a point when it was well maintained. A year goes by, it's been abandoned, but your star is still there. Hmmm. I'm having a hard time thinking of a better replacement.

@adg's spot on about the percent test coverage metric. That would require real compute time (and real money). That being said, it would be one of the better signals of project quality.

Regarding download counts: I do not think there is any real way of doing this. Even the GitHub [repos api](https://developer.github.com/v3/repos/) does not report this. You do get stargazer_count though.

**Comment by garyburd** [comment]
>> Number of includes in other public projects

> We already have this data. Seems like a no-brainer.

Search is partially ranked by number of imports now.

>> Number of pageviews for that project on GoDoc.org

>  We're not currently gathering this data, but it's probably worth doing.

The service stores a decaying count of page views with a halflife of one week to support the popular list on the home page.  It should be easy to connect this to search ranking.

**Comment by gdey** [comment]
> What are your goals for search ranking and how does ranking by these metrics support the goals?
> One goal might be to help developers find the best package in a some domain. Ranking by number of > imports supports this goal because each import is an indication that a developer found the package
> useful.

As you pointed out there are two goals for developers looking for packages.
1. Make it easier for developers to find a packages in a domain.
2. Make it easier for developers to evaluate the quality of packages in a domain.

Towards this end, the rankings based on imports and test help. As far as tests go, I think these need to be a bit liberal. The idea being is there at least one test, and it covers at least 5% percentage of the code. I don't think we should be expecting the percentage to be more then a single digit. I the the more important metric is the usage in other projects. 

For quality I think percentage of documentation for package and public  members is important. As well as the presence of examples, and testable examples is more important. As well as coming up with a ranking based on the output of go fmt, go vet, and other code linting tools. 

But there is another audience as well; that is the package authors themselves.
For them the goal is to provide feedback on how they can improve the quality of their projects. Having such a ranking system that is objective, and can show you what you need to do to improve your packages standing — I believe, will raise the base line quality of all packages.

>Ranking by recent activity does not necessarily support this goal. Activity on a high quality 
> mature package can be low. Activity on a buggy package can be high.

A part of the original proposal document that when towards building these requests was left out by mistake.

# Archive Expired Packages
An archive section to where libraries that have been inactive for a given period of time and not used by a number of other active projects. This would help keep the listing to only active projects. Inactive is defined as "Most recent commit > 365 days ago" and "number of imports < 10", and can be adjusted.

The 365 and 10 are just place holders and can be changed if needed. The idea here being, that we need to, also, group or filter packages; so that there isn't a overwhelming amount of choice. 

I would like to reference #90 as another issues that is trying to solve the getting too many packages to find the trees from the forest problem.

**Comment by garyburd** [comment]
Using search ranking as a stick to encourage package authors to improve metrics like test coverage may run counter to helping developers find the best package in a domain.  The best package may not have good test coverage. 

Displaying these metrics in search results can steer developers away from what may be the best package. 

The expired package idea looks promising. My intuition is that a conservative filter of 0 imports from outside of the repo and no commits in two years will filter out a lot of junk. I suggest writing a command line app to generate a list the packages to archive for a given import count and  last commit time and checking the results to see if they make sense.  The one complication is that gddo does not store the last commit time.  When I experimented with something similar a year or two ago, I used [this app](https://gist.github.com/garyburd/014b16f88b031e55f084) to save GitHub repo info to my local disk. I then fetched from my local disk instead of hitting GitHub repeatedly for the data.

**Comment by adg** [comment]
Note that this is related to #52

**Comment by adg** [comment]
And #172

**Comment by jbuberel** [comment]
This is a terrific idea, @garyburd:

>  I suggest writing a command line app to generate a list the packages to archive for a given import count and last commit time and checking the results to see if the filter is doing the right thing.

Implement the proposed filtering criteria, then test against the search results for a common term with a large result set, such as "web" or "sql" or "middleware". Generate side-by-side diff-able output with and without the filter.

Once we're confident that the filtering is "fair", then move onto changes in the ranking. Again, with the intent of being able to compare current vs proposed ranking so we can vet the diffs.

**Comment by garyburd** [comment]
@jbuberel The "archive expired packages" feature is a filter, not a ranking. To test the filter, we should look for false positives in the list of packages removed by the filter. 

I suggested writing a command line app to test the filter logic because the command line app is a small investment in time compared to implementing the feature in the service.  We don't know if the filter will have false positives or if the filter is worth the effort.  If the filter does not remove a significant percentage of the packages, it's not worth doing.

**Comment by jbuberel** [comment]
@garyburd Very much agreed.

**Comment by rafaeljusto** [comment]
@garyburd I started building the command line:
https://github.com/rafaeljusto/gddoexp

**Comment by garyburd** [comment]
@rafaeljusto Looks like a good start. I suggest using a local cache of GitHub repo data instead of hitting GitHub every time you execute the program. I used [this program](https://gist.github.com/garyburd/014b16f88b031e55f084) to create a local cache. You will need to edit the program to add your GitHub API id and secret.

**Comment by shurcooL** [comment]
> The expired package idea looks promising. My intuition is that a conservative filter of 0 imports from outside of the repo and no commits in two years will filter out a lot of junk.

I think that's a great idea. But, an observation, that will have false positives for commands or libraries meant to be used at go generate time, since they're typically imported from other packages in `// +build ignore` files. For similar reasons, it won't work for libraries that are meant to be used with OSes or architectures that godoc does not support/know about.

**Comment by rafaeljusto** [comment]
I ran the [tool](https://github.com/rafaeljusto/gddoexp) that checks the expired packages on a database dump from 2015-10-01. It analyzed **132277 packages** in **36h45m** due to Github rate limit policies. The results:

%      | Description
------ | ------------------
3.88 | not a Github project
2.75 | should be archived
0.63 | unexpected status code from Github

The unexpected status code from Github is probably some rate limit issues (403 Forbidden) that could be solved adjusting the token bucket values or analyzing the HTTP response headers from Github. The tool algorithm currents make two checks to identify if a package should be archived:

* No other packages reference the analyzed package (ImporterCount)
* Package wasn't modified is the last 2 years (Github response)

We also got 6 connections timeouts.

**Comment by garyburd** [comment]
@rafaeljusto Did you look at a sample of the packages flagged to archived to see if the list looks reasonable? Can you share the list of packages?

**Comment by jbuberel** [comment]
Indeed, the list of "should be archived" is crucial here. We need to sanity check that to ensure that no legitimately keep-worthy projects would be get the archive treatment :-) Can you @rafaeljusto pastebin or gist it for us?

**Comment by rafaeljusto** [comment]
I checked some of them to see if they were modified in the last 2 years. But I didn't check if they were referenced by other packages, I'm trusting in the gddo database information.

Here is the list of packages to archive:
https://gist.github.com/rafaeljusto/0ef14863b39c23517e0a

**Comment by garyburd** [comment]
The goal of the filter is to remove packages that are not useful. For example, we want to find packages that were abandoned before completion, inactive forks of other packages, etc.. Somebody should look at a sample of the packages found by this filter to see if there are any false positives.

**Comment by garyburd** [comment]
Packages with [score](https://godoc.org/github.com/golang/gddo/database#PackageInfo.Score) == 0.0 are not included in the search results.  Can you remove packages with score == 0.0 so we can see the packages identified by this filter only?

The package github.com/maximecaron/gcse/server is an example of a package with score == 0.  This package is not included in the search results because the repo is a fork with no modifications since the time of the fork.

**Comment by rafaeljusto** [comment]
Sure! I've created another [program](https://github.com/rafaeljusto/gddoexp/blob/master/cmd/score/score.go) that inform packages with score 0 (zero) from an input list. So, from the list of packages that should be archived we have:

%       | Description
------- | -----------------
67.30 | has no score
32.70 | has score

The list of packages with score that should be archived are here:
https://gist.github.com/rafaeljusto/d2795a100f4661b9b126

**Comment by garyburd** [comment]
In the sample I looked at, the repos are forks with at most two commits in the fork. In some cases the commits were merged upstream. A filter on forks with small number of commits close to the time of fork might be useful.

**Comment by rafaeljusto** [comment]
Working on it. =)

**Comment by rafaeljusto** [comment]
I've created a new filter that checks for forks with maximum of 2 commits in the week after the fork date, I called then "fast forks".

On the list of scored packages that should be archived, when applying this filter we got:

%       | Description
------- | --------------
51.05 | fast fork
48.95 | not fast fork

The list of packages after applying this 2 filters can be found bellow:
https://gist.github.com/rafaeljusto/3131e9e43c905d2e0808

**Comment by jbuberel** [comment]
I just spot-checked about 50 items from the new list, and I didn't see any false-positives (project that would have been archived but should not have been). So far, LGTM.

**Comment by garyburd** [comment]
I suggest using the "fast fork" filter as an alternative to the two year filter in the original program.  The purpose of my suggestion is to increase the number of packages hidden by the filter.

**Comment by jbuberel** [comment]
Sounds like we're in general agreement that the new fast-fork filter is working well as an identifier of packages that should be considered "archived" and therefore not displayed in search results.

Using that filtered set as the base, it make sense to begin experimenting with the rankings (the primary goal of this proposal).  Given that gddo already takes import counts into account, how about an experiment to apply the stored page view counts for a small set of common search terms ("sql" and "middleway"), with outputs that allow us to diff/compare the ordering of:

* filtered + ranked by imports
* filtered + ranked by imports + ranked by pageviews

**Comment by rafaeljusto** [comment]
I ran the tool again now replacing the 2 years condition for the fast fork. It analyzed **132277 packages** in **4h0m42s** (we got many cache hits). The results:

%       | Description
------- | ----------------------------
3.88   | not a Github project
14.14 | should be archived
1.47   | unexpected status code from Github

We also got 9 connections timeouts.

The new list of packages that could be archived are bellow:
https://gist.github.com/rafaeljusto/db8318b69efb10f622aa

We increased the packages to archive in **11.39%** comparing with the first result. I think we could apply both rules: we archive if the package is a fast fork or has more than two years with no changes, already considering that there are no other packages referencing it.

There are other cases where we got a 404 from Github API that we could also archive, but this are only a few cases. I still need to [work](https://github.com/rafaeljusto/gddoexp/issues/2) on the tool to avoid rate limit and decrease the "unexpected status code from Github" percentage.

**PS:** I will be offline for a week (hello vacations!)

**Comment by garyburd** [comment]
The current filter results are good in that there are no false positives, but the filters may not be worth implementing given the small number of packages found.  

Consider this example: https://godoc.org/?q=redis . The proposed filters do little to reduce the noise in these search results.

The server is currently written to filter packages using a score computed at a point in time. Filters that fit this model are easy to implement.  An example is the current "dead fork" filter.  If the repo push time is before the repo creation time and the repo is a fork, then the package is not included in the search index. 

Filtering based on time since last update to the repo does not fit the point in time score. It's possible to implement a time based filter. I personally would want to know that it makes a difference before doing the work.

I suggest saying that a package is "removed from search results" or "not indexed" instead of "archived".  The server is not going to store the identified packages in an archive.

**Comment by garyburd** [comment]
> We increased the packages to archive in 11.39%

Nice!  Removing 1 of 10 results from the search results is noticeable.

> 404 from Github API that we could also archive

The crawler will remove these packages from the database. This is not a concern for search.

> avoid rate limit and decrease the "unexpected status code from Github" percentage

The tool that I shared earlier in the thread does breach rate limits or get unexpected status codes.  I'll clean it up and post it so that others can use it easily.

**Comment by carlisia** [comment]
This writeup is relevant for this discussion: https://github.com/mikeal/go-stats/blob/master/README.md

**Comment by margaery** [comment]
@carlisia I read the document, but I do not understand how it's relevant to this issue. Can you explain?

**Comment by jamra** [comment]
@carlisia I don't know why package count should be compared with other communities. It shouldn't matter. Some Go projects on github divide their package into many small packages. That is the Go way to do it and helps make things go gettable.

In terms of ranking packages on godoc, you bring up a good point: We can use an "imported by" metric to rank packages. That could remove some of the noise added by some of these sub packages.
